paasta_tools/kubernetes_tools.py

```diff
@@ -0,0 +1,581 @@
+                        ),
+                        "restartPolicy": "Always",
+                        "volumes": self.get_pod_volumes(docker_volumes) + [{
+                            "emptyDir": {},
+                            "name": self.get_sanitised_volume_name('/var/spool/hacheck'),
+                        }],
+                    },
+                },
+            },
+        }
+
+        config_hash = get_config_hash(
+            self.sanitize_for_config_hash(complete_config),
+            force_bounce=self.get_force_bounce(),
+        )
+        complete_config['metadata']['labels']['config_sha'] = config_hash
+        complete_config['spec']['template']['metadata']['labels']['config_sha'] = config_hash
+        complete_config['metadata']['labels']['git_sha'] = code_sha
+        complete_config['spec']['template']['metadata']['labels']['git_sha'] = code_sha
+
+        log.debug("Complete configuration for instance is: %s", complete_config)
+        return complete_config
+
+    def sanitize_for_config_hash(
+        self,
+        config: Dict,
+    ) -> Dict[str, Any]:
+        """Removes some data from config to make it suitable for
+        calculation of config hash.
+
+        :param config: complete_config hash to sanitize
+        :returns: sanitized copy of complete_config hash
+        """
+        ahash = {key: copy.deepcopy(value) for key, value in config.items() if key not in CONFIG_HASH_BLACKLIST}
+        spec = ahash['spec']
+        ahash['spec'] = {key: copy.deepcopy(value) for key, value in spec.items() if key not in CONFIG_HASH_BLACKLIST}
+        return ahash
+
+    def get_bounce_margin_factor(self) -> float:
+        return self.config_dict.get('bounce_margin_factor', 1.0)
+
+
+def read_all_registrations_for_service_instance(
+    service: str,
+    instance: str,
+    cluster: Optional[str]=None,
+    soa_dir: str=DEFAULT_SOA_DIR,
+) -> List[str]:
+    """Retreive all registrations as fully specified name.instance pairs
+    for a particular service instance.
+
+    For example, the 'main' paasta instance of the 'test' service may register
+    in the 'test.main' namespace as well as the 'other_svc.main' namespace.
+
+    If one is not defined in the config file, returns a list containing
+    name.instance instead.
+    """
+    if not cluster:
+        cluster = load_system_paasta_config().get_cluster()
+
+    kubernetes_service_config = load_kubernetes_service_config(
+        service=service,
+        instance=instance,
+        cluster=cluster,
+        load_deployments=False,
+        soa_dir=soa_dir,
+    )
+    return kubernetes_service_config.get_registrations()
+
+
+def get_kubernetes_services_running_here() -> List[KubeService]:
+    services = []
+    pods = requests.get('http://127.0.0.1:10255/pods').json()
+    for pod in pods['items']:
+        if pod['status']['phase'] != 'Running' or pod['metadata']['namespace'] != 'paasta':
```

Comment:
I didn't see the namespace being set anywhere in format_kubernetes_app_dict?

Reply:
It seems to be a separate param to the API so we set it in the `create_` call later